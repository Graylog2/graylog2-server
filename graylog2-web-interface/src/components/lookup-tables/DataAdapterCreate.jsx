import PropTypes from 'prop-types';
import React from 'react';
import naturalSort from 'javascript-natural-sort';

import { Row, Col } from 'react-bootstrap';
import { Input } from 'components/bootstrap';
import { Select } from 'components/common';
import { DataAdapterForm } from 'components/lookup-tables';
import { PluginStore } from 'graylog-web-plugin/plugin';
import ObjectUtils from 'util/ObjectUtils';

class DataAdapterCreate extends React.Component {
  static propTypes = {
    saved: PropTypes.func.isRequired,
    types: PropTypes.object.isRequired,
    validate: PropTypes.func,
    validationErrors: PropTypes.object,
  };

  static defaultProps = {
    validate: null,
    validationErrors: {},
  };

  state = {
    dataAdapter: undefined,
    type: undefined,
  };

  _onTypeSelect = (adapterType) => {
    this.setState({
      type: adapterType,
      dataAdapter: {
        id: null,
        title: '',
        name: '',
        description: '',
        config: ObjectUtils.clone(this.props.types[adapterType].default_config),
      },
    });
  };

  render() {
    const adapterPlugins = {};
    PluginStore.exports('lookupTableAdapters').forEach((p) => {
      adapterPlugins[p.type] = p;
    });

    const sortedAdapters = Object.keys(this.props.types).map((key) => {
      const type = this.props.types[key];
      if (adapterPlugins[type.type] === undefined) {
        console.error(`Plugin component for data adapter type ${type.type} is missing - invalid or missing plugin?`);
        return { value: type.type, disabled: true, label: `${type.type} - missing or invalid plugin` };
      }
      return { value: type.type, label: adapterPlugins[type.type].displayName };
    }).sort((a, b) => naturalSort(a.label.toLowerCase(), b.label.toLowerCase()));

    return (<div>
      <Row className="content">
        <Col lg={8}>
          <form className="form form-horizontal" onSubmit={() => {}}>
            <Input id="data-adapter-type-select"
                   label="Data Adapter Type"
                   required
                   autoFocus
                   help="The type of data adapter to configure."
                   labelClassName="col-sm-3"
                   wrapperClassName="col-sm-9">
              <Select placeholder="Select Data Adapter Type"
                      clearable={false}
                      options={sortedAdapters}
                      matchProp="value"
                      onChange={this._onTypeSelect}
                      value={null} />
            </Input>
          </form>
        </Col>
      </Row>
      {this.state.dataAdapter && (
        <Row className="content">
          <Col lg={12}>
            <h3>Configure Adapter</h3>
            <DataAdapterForm dataAdapter={this.state.dataAdapter}
                             type={this.state.type}
                             create
                             validate={this.props.validate}
                             validationErrors={this.props.validationErrors}
                             saved={this.props.saved} />
          </Col>
        </Row>
      )}
    </div>);
  }
}


export default DataAdapterCreate;
