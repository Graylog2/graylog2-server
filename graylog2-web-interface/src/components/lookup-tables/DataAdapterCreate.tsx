/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';

import type {
  LookupTableAdapter,
  validationErrorsType,
  LUTTypesAPIResponse,
  LUTTypesType,
} from 'logic/lookup-tables/types';
import usePluginEntities from 'hooks/usePluginEntities';
import { Select } from 'components/common';
import { Row, Col, Input } from 'components/bootstrap';
import { DataAdapterForm } from 'components/lookup-tables';

import type { DataAdapterPluginType } from './types';

const INIT_DATA_ADAPTER: LookupTableAdapter = {
  id: null,
  title: '',
  name: '',
  description: '',
  config: { type: 'none' },

};

type OptionType = { value: string, label: string, disabled?: boolean };

type Props = {
  saved: () => void,
  types: LUTTypesAPIResponse,
  validate: () => void,
  validationErrors: validationErrorsType,
};

const DataAdapterCreate = ({ saved, types, validate, validationErrors }: Props) => {
  const [type, setType] = React.useState<string>(null);
  const dataAdapterPlugins = usePluginEntities('lookupTableAdapters');

  const plugins = React.useMemo(() => (
    dataAdapterPlugins.reduce((acc: any, plugin: DataAdapterPluginType) => {
      acc[plugin.type] = plugin;

      return acc;
    }, {})
  ), [dataAdapterPlugins]);

  const adapterTypes = React.useMemo(() => (
    Object.values(types)
      .map((inType: LUTTypesType) => {
        if (!plugins[inType.type]) {
          // eslint-disable-next-line no-console
          console.error(`${inType.type} - missing or invalid plugin`);

          return {
            value: inType.type,
            disabled: true,
            label: `${inType.type} - missing or invalid plugin`,
          };
        }

        return { value: inType.type, label: plugins[inType.type].displayName };
      }).sort((a: OptionType, b: OptionType) => {
        if (a.label.toLowerCase() > b.label.toLowerCase()) return 1;
        if (a.label.toLowerCase() < b.label.toLowerCase()) return -1;

        return 0;
      })
  ), [types, plugins]);

  const dataAdapter = React.useMemo(() => {
    if (type) {
      return {
        ...INIT_DATA_ADAPTER,
        config: { ...types[type]?.default_config },
      };
    }

    return null;
  }, [type, types]);

  const handleSelect = (selectedType: string) => {
    setType(selectedType);
  };

  return (
    <>
      <Row className="content">
        <Col lg={8}>
          <form className="form form-horizontal" onSubmit={() => {}}>
            <Input id="data-adapter-type-select"
                   label="Data Adapter Type"
                   required
                   autoFocus
                   help="The type of data adapter to configure."
                   labelClassName="col-sm-3"
                   wrapperClassName="col-sm-9">
              <Select placeholder="Select Data Adapter Type"
                      clearable={false}
                      options={adapterTypes}
                      matchProp="label"
                      onChange={handleSelect}
                      value={null} />
            </Input>
          </form>
        </Col>
      </Row>
      {dataAdapter && (
        <Row className="content">
          <Col lg={12}>
            <DataAdapterForm dataAdapter={dataAdapter}
                             type={type}
                             create
                             title="Configure Adapter"
                             validate={validate}
                             validationErrors={validationErrors}
                             saved={saved} />
          </Col>
        </Row>
      )}
    </>
  );
};

export default DataAdapterCreate;
