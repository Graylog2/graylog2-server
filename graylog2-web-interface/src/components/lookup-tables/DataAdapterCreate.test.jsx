import React from 'react';
import renderer from 'react-test-renderer';

import mockComponent from 'helpers/mocking/MockComponent';

import DataAdapterCreate from './DataAdapterCreate';

jest.mock('components/common', () => ({
  Select: mockComponent('SelectMock'),
}));

jest.mock('components/bootstrap', () => ({
  Input: mockComponent('InputMock'),
}));

jest.mock('components/lookup-tables', () => ({
  DataAdapterForm: mockComponent('DataAdapterFormMock'),
}));

jest.mock('graylog-web-plugin/plugin', () => ({
  PluginStore: {
    exports: () => ([
      {
        type: 'someType',
        displayName: 'Some Mocked Data Adapter Type',
      },
    ]),
  },
}));

describe('<DataAdapterCreate />', () => {
  it('should render with empty parameters', () => {
    const callback = () => {};
    const types = {};
    const wrapper = renderer.create(<DataAdapterCreate saved={callback} types={types} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render for types with defined frontend components', () => {
    const callback = () => {};
    const types = {
      someType: {
        type: 'someType',
      },
    };
    const wrapper = renderer.create(<DataAdapterCreate saved={callback} types={types} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  describe('with mocked console.error', () => {
    const consoleError = console.error;

    beforeAll(() => {
      console.error = jest.fn();
    });

    afterAll(() => {
      console.error = consoleError;
    });

    it('should render for types without defined frontend components', () => {
      const callback = () => {};
      const types = {
        unknownType: {
          type: 'unknownType',
        },
      };
      const wrapper = renderer.create(<DataAdapterCreate saved={callback} types={types} />);
      expect(wrapper.toJSON()).toMatchSnapshot();
      expect(console.error.mock.calls.length).toBe(1);
    });
  });
});
