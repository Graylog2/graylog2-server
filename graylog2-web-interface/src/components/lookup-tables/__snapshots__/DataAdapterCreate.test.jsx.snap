// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataAdapterCreate /> should render for types with defined frontend components 1`] = `
<div>
  <div
    className="content row"
  >
    <div
      className="col-lg-8"
    >
      <form
        className="form form-horizontal"
        onSubmit={[Function]}
      >
        <input-mock
          autoFocus={true}
          help="The type of data adapter to configure."
          id="data-adapter-type-select"
          label="Data Adapter Type"
          labelClassName="col-sm-3"
          required={true}
          wrapperClassName="col-sm-9"
        >
          <select-mock
            clearable={false}
            matchProp="value"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "Some Mocked Data Adapter Type",
                  "value": "someType",
                },
              ]
            }
            placeholder="Select Data Adapter Type"
            value={null}
          />
        </input-mock>
      </form>
    </div>
  </div>
</div>
`;

exports[`<DataAdapterCreate /> should render with empty parameters 1`] = `
<div>
  <div
    className="content row"
  >
    <div
      className="col-lg-8"
    >
      <form
        className="form form-horizontal"
        onSubmit={[Function]}
      >
        <input-mock
          autoFocus={true}
          help="The type of data adapter to configure."
          id="data-adapter-type-select"
          label="Data Adapter Type"
          labelClassName="col-sm-3"
          required={true}
          wrapperClassName="col-sm-9"
        >
          <select-mock
            clearable={false}
            matchProp="value"
            onChange={[Function]}
            options={Array []}
            placeholder="Select Data Adapter Type"
            value={null}
          />
        </input-mock>
      </form>
    </div>
  </div>
</div>
`;

exports[`<DataAdapterCreate /> with mocked console.error should render for types without defined frontend components 1`] = `
<div>
  <div
    className="content row"
  >
    <div
      className="col-lg-8"
    >
      <form
        className="form form-horizontal"
        onSubmit={[Function]}
      >
        <input-mock
          autoFocus={true}
          help="The type of data adapter to configure."
          id="data-adapter-type-select"
          label="Data Adapter Type"
          labelClassName="col-sm-3"
          required={true}
          wrapperClassName="col-sm-9"
        >
          <select-mock
            clearable={false}
            matchProp="value"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "disabled": true,
                  "label": "unknownType - missing or invalid plugin",
                  "value": "unknownType",
                },
              ]
            }
            placeholder="Select Data Adapter Type"
            value={null}
          />
        </input-mock>
      </form>
    </div>
  </div>
</div>
`;
